velox/exec/NestedLoopJoinProbe.cpp

```diff
@@ -80,15 +122,224 @@ void NestedLoopJoinProbe::addInput(RowVectorPtr input) {
     child->loadedVector();
   }
   input_ = std::move(input);
+  VELOX_CHECK_EQ(buildIndex_, 0);
+  if (needsProbeMismatch(joinType_)) {
+    probeMatched_.resizeFill(input_->size(), false);
+  }
 }
 
 RowVectorPtr NestedLoopJoinProbe::getOutput() {
-  if (input_ == nullptr) {
+  if (isFinished()) {
     return nullptr;
   }
+  RowVectorPtr output{nullptr};
+  while (output == nullptr) {
+    if (lastProbe_) {
+      VELOX_CHECK(processingBuildMismatch());
+      VELOX_CHECK_NULL(input_);
 
-  const auto inputSize = input_->size();
+      while (output == nullptr && !hasProbedAllBuildData()) {
+        output = getMismatchedOutput(
+            buildVectors_.value()[buildIndex_],
+            buildMatched_[buildIndex_],
+            buildOutMapping_,
+            buildProjections_,
+            identityProjections_);
+        ++buildIndex_;
+      }
+      if (hasProbedAllBuildData()) {
+        setState(ProbeOperatorState::kFinish);
+      }
+      break;
+    }
+
+    if (input_ == nullptr) {
```

Comment:
I am curious when will we run into null input here?

Reply:
On `driver`'s first `getOutput()` call to the operator:
https://github.com/facebookincubator/velox/blob/333b870cf2d4668c4b4dfff29c1450316081b65d/velox/exec/Driver.cpp#L399

When processing a pipeline, `driver` does the following to operators, in reverse order(from sink to source):
```
// this is simplified
if !currentOperator.isBlocked()
    if nextOperator.needsInput()
        output = currentOperator.getOutput()
        if output
            nextOperator.addInput(output)
            currentOperator = nextOperator
        else
            check if currentOperator is blocked or finished
            if not, currentOperator = prevOperator
```

As `driver` always tries to find the last operator that can emit output, `getOutput` is called to see if it can output something before input is added to the operator, the `if` is for this case.